metric.go

```diff
@@ -4,207 +4,106 @@ import (
 	"bytes"
 	"crypto/md5"
 	"encoding/binary"
-	"encoding/json"
 	"errors"
 	"fmt"
 	"sort"
 )
 
-var errInvalidIntervalzero = errors.New("interval cannot be 0")
 var errInvalidOrgIdzero = errors.New("org-id cannot be 0")
 var errInvalidEmptyName = errors.New("name cannot be empty")
-var errInvalidEmptyMetric = errors.New("metric cannot be empty")
 var errInvalidMtype = errors.New("invalid mtype")
 var errInvalidTagFormat = errors.New("invalid tag format")
 
-type PartitionedMetric interface {
-	Validate() error
-	SetId()
-	// return a []byte key comprised of the metric's OrgId
-	// accepts an input []byte to allow callers to re-use
-	// buffers to reduce memory allocations
-	KeyByOrgId([]byte) []byte
-	// return a []byte key comprised of the metric's Name
-	// accepts an input []byte to allow callers to re-use
-	// buffers to reduce memory allocations
-	KeyBySeries([]byte) []byte
-}
+var bPool = NewBufferPool()
 
 //go:generate msgp
 
-// MetricData contains all metric metadata (some as fields, some as tags) and a datapoint
-type MetricData struct {
-	Id       string   `json:"id"`
-	OrgId    int      `json:"org_id"`
-	Name     string   `json:"name"`
-	Metric   string   `json:"metric"`
-	Interval int      `json:"interval"`
-	Value    float64  `json:"value"`
-	Unit     string   `json:"unit"`
-	Time     int64    `json:"time"`
-	Mtype    string   `json:"mtype"`
-	Tags     []string `json:"tags" elastic:"type:string,index:not_analyzed"`
+//msgp:tuple Point
+type Point struct {
+	Ts  uint32
+	Val float64
 }
 
-func (m *MetricData) Validate() error {
-	if m.OrgId == 0 {
-		return errInvalidOrgIdzero
-	}
-	if m.Interval == 0 {
-		return errInvalidIntervalzero
-	}
-	if m.Name == "" {
-		return errInvalidEmptyName
-	}
-	if m.Metric == "" {
-		return errInvalidEmptyMetric
-	}
-	if m.Mtype == "" || (m.Mtype != "gauge" && m.Mtype != "rate" && m.Mtype != "count" && m.Mtype != "counter" && m.Mtype != "timestamp") {
-		return errInvalidMtype
-	}
-	if !ValidateTags(m.Tags) {
-		return errInvalidTagFormat
-	}
-	return nil
+func (m Point) Timestamp() uint32 {
+	return m.Ts
 }
 
-func (m *MetricData) KeyByOrgId(b []byte) []byte {
-	if cap(b)-len(b) < 4 {
-		// not enough unused space in the slice so we need to grow it.
-		newBuf := make([]byte, len(b), len(b)+4)
-		copy(newBuf, b)
-		b = newBuf
-	}
-	// PutUint32 writes directly to the slice rather then appending.
-	// so we need to set the length to 4 more bytes then it currently is.
-	b = b[:len(b)+4]
-	binary.LittleEndian.PutUint32(b[len(b)-4:], uint32(m.OrgId))
-	return b
+func (m Point) Value() float64 {
+	return m.Val
 }
 
-func (m *MetricData) KeyBySeries(b []byte) []byte {
-	b = append(b, []byte(m.Name)...)
-	return b
+type MetricPoint struct {
+	Id string
```

Comment:
I thought we agreed in #199 that Id's should be stored as a `[16]byte` (and a `uint32` for orgId when needed)